### PR TITLE
DA_Week4_복습과제_손소현

### PR DESCRIPTION
**Pull request by sonso1598**

### CODE REVIEW

**Comment by yegenuine** [comment]
안녕하세요 소현님!
이번 복습과제 문제 양이 정말 많았는데요, 열심히 풀어주셔서 감사합니다.
작성해주신 코드 중 답안지와 약간 다르거나 추가로 알아두시면 좋을 것 같은 사항들 몇가지만 피드백에 적어봤어요. 특히 문제 앞부분에서 문제가 발생했던 스태킹, CatBoost 파트 위주로 확인해주시면 될 것 같습니다.

1. 데이터에서 피쳐값과 타겟값을 분리하여 각각 X_wine, y_wine으로 저장하시오. (target은 quality)
X_wine = wine_data.drop(['quality'], axis=1)
y_wine = wine_data['quality']
타겟값이 들어있는 변수명을 알고 있기 때문에 이렇게 분리하는 방법도 알아두시면 좋을 것 같습니다.

2. 피쳐값 중 카테고리 변수가 1가지 있다. 이에 대해 Encoding을 진행하시오.
X_wine 데이터의 카테고리 변수를 Encoding할 때 get_dummies() 함수로도 가능한 점 알아두시면 좋을 것 같아요.
X_wine['type'] = pd.get_dummies(X_wine['type'])

3. 사이킷런 래퍼 클래스 LightGBM와 GridSearchCV를 사용해 최적의 하이퍼 파라미터 조합과 그 하이퍼 파라미터의 예측 정확도를 구하시오.
깃허브 상으로 보면 오류가 발생한 것으로 보여 답안지 코드를 첨부합니다!
from sklearn.model_selection import GridSearchCV
from lightgbm import LGBMClassifier

params = {
    'n_estimators' : [100,200,300,400,500],
    'learning_rate' : [0.01,0.05, 0.1, 0.15, 0.2],
    'max_depth' : [3,4,5,6,7]
}

lgbm_wrapper = LGBMClassifier()
lgbm_wrapper.fit(X_train, y_train)
grid_cv = GridSearchCV(lgbm_wrapper, param_grid=params, cv=2, verbose=1)
grid_cv.fit(X_train,y_train)
print('최적 하이퍼 하라미터:', grid_cv.best_params_)
print('최고 예측 정확도:', grid_cv.best_score_)

4. CV 세트 기반의 스태킹을 진행하려 한다. 개별 기반 모델들이 교차 검증을 통해 최종 메타 모델이 사용할 학습용, 테스트용 데이터를 생성하기 위한 함수를 작성하시오.
get_stacking_base_datasets() 함수 작성하실 때 교재에서 사용했던 데이터와 달리 데이터프레임 타입의 데이터에 사용해야 하기 때문에
X_tr = X_train_n.iloc[train_index] 
y_tr = y_train_n.iloc[train_index] 
X_te = X_train_n.iloc[valid_index]
이렇게 iloc을 사용해주셔야 합니다.
나머지 코드에는 문제가 없어보이기 때문에 이 부분 수정하시면 스태킹 파트 잘 진행될 것 같습니다.

5. CatBoost를 진행할 수 있도록 범주형 변수를 처리하시오.
Pool()을 이용하여 범주형 변수를 처리할 때,
train_pool = Pool(data=X_train, label=y_train, cat_features = ['cap-shape', ... , 'habitat'])
이렇게 cat_features에 범주형 변수명들을 같이 넣어주셔야 합니다.
나머지 코드에는 문제가 없어보이기 때문에 이 부분 수정하시면 CABT파트 잘 진행될 것 같습니다.

고생 많으셨습니다. 감사합니다!!
